packages/core/src/config/parse.ts

```diff
@@ -1682,64 +1783,47 @@ const parseAndValidateChugSplashConfig = async (
       )
     }
 
-    const { externalProxy, externalProxyType, constructorArgs } =
-      userContractConfig
-
     // Change the `contract` fields to be a fully qualified name. This ensures that it's easy for the
     // executor to create the `CanonicalConfigArtifacts` when it eventually compiles the canonical
     // config.
-    const { sourceName, contractName } = readContractArtifact(
-      artifactPaths[referenceName].contractArtifactPath,
-      integration
-    )
+    const artifact = cachedArtifacts[referenceName]
+    const { sourceName, contractName } = artifact
     const contractFullyQualifiedName = `${sourceName}:${contractName}`
 
-    // Set the proxy address to the user-defined value if it exists, otherwise set it to the default proxy
-    // used by ChugSplash.
-    const proxy =
-      externalProxy ||
-      getDefaultProxyAddress(userConfig.options.projectName, referenceName)
-
-    let proxyType: ProxyType
-    if (externalProxyType) {
-      proxyType = externalProxyType
+    let kind: ProxyType
+    if (userContractConfig.kind) {
+      kind = userContractConfig.kind
     } else {
-      proxyType = 'internal-default'
+      kind = 'internal-default'
     }
 
-    const { output: compilerOutput } = readBuildInfo(
-      artifactPaths[referenceName].buildInfoPath
-    )
+    const compilerOutput = cachedCompilerOutput[referenceName]
     const storageLayout =
       compilerOutput.contracts[sourceName][contractName].storageLayout
 
-    const parsedVariables = parseContractVariables(
-      JSON.parse(
-        Handlebars.compile(JSON.stringify(userContractConfig))({
-          ...contracts,
-        })
-      ),
-      storageLayout,
-      compilerOutput,
-      cre
-    )
-
-    const args = parseAndValidateConstructorArgs(
-      constructorArgs ?? {},
-      referenceName,
-      compilerOutput.contracts[sourceName][contractName].abi,
-      cre
-    )
+    // Only run the parsing logic if the contract is proxied
+    let parsedVariables
+    if (userContractConfig.kind !== 'no-proxy') {
+      const variables = parseContractVariables(
+        JSON.parse(
+          Handlebars.compile(JSON.stringify(userContractConfig))({
+            ...contracts,
+          })
```

Comment:
Instead of calling `Handlebars.compile` inside this for-loop, I think we should call it before the for-loop. It's only necessary to resolve the contract references once